### PR TITLE
Framework: add <Sidebar> component under Layout

### PR DESCRIPTION
**Pull request by mtias**
This component will take the place of all the custom sidebar elements we are using, and consolidate all the sub-components used to construct the sidebar (sidebar menu, sidebar item, sidebar heading, etc) as stateless pure components.

Part of this work will also result in consolidating the Sass code for the sidebar, and getting rid of `stylesheets/layout/_sidebar`.

### CODE REVIEW

File: client/layout/sidebar/index.jsx [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+	displayName: 'Sidebar',
+
+	propTypes: {
+		className: React.PropTypes.string,
+		onClick: React.PropTypes.func
+	},
+
+	render: function() {
+		return (
+			<ul className="sidebar" onClick={ this.props.onClick }>
```
Comment (aduth) [inline client/layout/sidebar/index.jsx]: I find it a bit odd that we render this as a `<ul />`, especially when many of its children aren't `<li>` but rather header elements. Should we use [`<nav />`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/nav) instead or consider wrapping the headers in `<li />`?

File: client/layout/sidebar/menu.jsx [hunk 2]
```diff
@@ -0,0 +1,10 @@
+/**
+ * External dependencies
+ */
+import React from 'react';
+
+const SidebarMenu = ( { children } ) => (
+	<li className="sidebar__menu">{ children }</li>
```
Comment (aduth) [inline client/layout/sidebar/menu.jsx]: I notice that most all usages of `<SidebarMenu />` render a `<ul>` and optionally a header. I wonder if we could make this a bit easier by removing the concern for rendering the list, i.e.:

```jsx
// In usage:
<SidebarMenu heading={ this.translate( 'My Section' ) }>
	<MyMenuItem />
	<MyMenuItem />
</SidebarMenu>

// Implemented as:
const SidebarMenu = ( { heading, children } ) => (
	<li className="sidebar__menu">
		{ heading && <SidebarHeading>{ heading }</SidebarHeading> }
		<ul>
			{ React.Children.map( children, ( child ) => {
				<li>{ child }</li>
			} ) }
		</ul>
	</li>
);
```

File: client/my-sites/upgrades/cart/secondary-cart.jsx [hunk 3]
```diff
@@ -19,18 +20,18 @@ var SecondaryCart = React.createClass( {
...
-					selectedSite={ selectedSite }
-					cart={ this.props.cart } />
-
-				<CartBody
-					cart={ this.props.cart }
-					selectedSite={ selectedSite }
-					showCoupon={ true } />
-			</div>
+			<Sidebar>
+				<div className="secondary-cart">
```
Comment (aduth) [inline client/my-sites/upgrades/cart/secondary-cart.jsx]: Maybe support `className` extending on `<Sidebar />`?

File: client/layout/sidebar/index.jsx [hunk 4]
```diff
@@ -0,0 +1,21 @@
+/**
+ * External dependencies
+ */
+import React from 'react';
+
+export default React.createClass( {
+	displayName: 'Sidebar',
+
+	propTypes: {
+		className: React.PropTypes.string,
```
Comment (aduth) [inline client/layout/sidebar/index.jsx]: Not used. [Related?](https://github.com/Automattic/wp-calypso/pull/2272/files#r49600502)

File: assets/stylesheets/layout/_main.scss [hunk 5]
```diff
@@ -291,10 +291,10 @@ $sidebar-width-min: 228px;
 	transition: all 0.15s ease-in-out;
```
Comment (aduth) [inline assets/stylesheets/layout/_main.scss]: There's still [a reference to `.wpcom-sidebar`](https://github.com/Automattic/wp-calypso/blob/ee7daccb1aa7922d1ba6b97c3df1258dd8502b3e/assets/stylesheets/layout/_main.scss#L318-L320) in this file, which I assume should be removed/updated?

File: client/layout/sidebar/menu.jsx (hunk 2 above)
Comment (mtias) [inline client/layout/sidebar/menu.jsx]: Makes sense. It would also allow us to eventually drop that markup if we want without having to touch the files.

Comment (mtias) [inline client/layout/sidebar/menu.jsx]: I'm going to finish this one once I create a centralized `sidebar/item.jsx`.

File: client/my-sites/upgrades/cart/secondary-cart.jsx (hunk 3 above)
Comment (mtias) [inline client/my-sites/upgrades/cart/secondary-cart.jsx]: I forgot to use the prop.

File: client/layout/sidebar/index.jsx (hunk 1 above)
Comment (mtias) [inline client/layout/sidebar/index.jsx]: I agree, but wanted to get the components in place first without changing the markup. I will loop back to it once I create sidebar-item.

**Comment by mtias** [comment]
@aduth good to go? I'll work on the sidebar-item and markup cleanup next.

**Comment by gwwar** [comment]
:+1: Tested in Chrome, FF, IE 11, Edge

**Comment by aduth** [comment]
Looks good :+1:

File: client/layout/sidebar/menu.jsx (hunk 2 above)
Comment (ockham) [inline client/layout/sidebar/menu.jsx]: > I'm going to finish this one once I create a centralized `sidebar/item.jsx`.

You might be able to reduce some of `client/my-sites/sidebar/sidebar-menu-item.jsx`.

Comment (mtias) [inline client/layout/sidebar/menu.jsx]: @ockham working a bit on this one here: https://github.com/Automattic/wp-calypso/pull/2419

Comment (ockham) [inline client/layout/sidebar/menu.jsx]: @mtias Thanks for the pointer!

**Pull request merged**